dohproxy/httpproxy.py

```diff
@@ -47,6 +47,11 @@ async def doh1handler(request):
             return aiohttp.web.Response(status=400, body=e.body())
     elif request.method == 'POST':
         body = await request.content.read()
+        # size without length indicator
+        if len(body) - 2 > constants.DOH_MAX_QUERY_SIZE:
+            return aiohttp.web.Response(
+                status=400, body=b'Size limit exceeded'
+            )
```

Comment:
I am thinking we can detect this even earlier by checking [Content-Length](https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.BaseRequest.content_length). This way we don't even need to read the body.
If None, [411](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/411) can be returned, if greater than DOH_MAX_QUERY_SIZE than [413](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413) maybe?

Reply:
* Does ``aiohttp`` verify content length and ensure that a client cannot send more data than it indicates in content length?
* RFC 8484 does mention chunked POST requests. A chunked request does not have content length header. If chunking is not supported then ``411 Length Required`` is a good idea.

Reply:
I did not get into the details of aiohttp implementation, but it does some validation (at least that content-length is set for non-chunked) as far as I can tell in the [Payload Parser](https://github.com/aio-libs/aiohttp/blob/9c1080647a2ea07695b54bc9838118e4ad825d09/aiohttp/http_parser.py#L522). It does not seem to validate length after reading, but I suppose an early check can help preventing reading payload in the first place, the second check would ensure we did not get lied to?
Given that the read is a lazy read, if you can prevent reading this data in the first place, why not do so?

> RFC 8484 does mention chunked POST requests.
I am failing to find it, mind linking to the section?

As much as supporting chunked POST requests, I don't see any good reason to do so, so I would not bother really. This is supposedly only receiving small DNS message.